src/facades/TradingWalletFacade.js

```diff
@@ -0,0 +1,122 @@
+const log = require('../logger')
+const { QuantityNotEnoughError } = require('../utils').errors
+const { TransactionLibBuilder } = require('../factories')
+
+/**
+ * Class representing a facade of tradingWallet utilities.
+ */
+
+class TradingWalletFacade {
+  constructor(tradingWalletService, erc20TokenService, transactionLib = TransactionLibBuilder.build(), logger = log) {
+    if (!logger) {
+      throw new TypeError(`Invalid "logger" value: ${logger}`)
+    }
+    this.log = logger.child({ module: this.constructor.name })
+
+    if (!tradingWalletService) {
+      const errorMessage = `Invalid "tradingWalletService" value: ${tradingWalletService}`
+      throw new TypeError(errorMessage)
+    }
+    this.tradingWalletService = tradingWalletService
+
+    if (!erc20TokenService) {
+      const errorMessage = `Invalid "erc20TokenService" value: ${erc20TokenService}`
+      throw new TypeError(errorMessage)
+    }
+    this.erc20TokenService = erc20TokenService
+
+    if (!transactionLib) {
+      const errorMessage = `Invalid "transactionLib" value: ${transactionLib}`
+      throw new TypeError(errorMessage)
+    }
+    this.transactionLib = transactionLib
+  }
+
+  async depositTokenAsync(personalWalletAddress, tradingWalletAddress, quantity, tokenAddress, privateKey) {
```

Comment:
Unfortunately this function for how it is done does not work in case of `allowance === 0`.
The `tradingWalletService.depositTokenAsync` will use `gasEstimation` under the hood and will fail until that the `approve` state transition takes effect.

We could:
- move `TransactionLib getTransactionObject/getGasEstimation` function to public
- use the `TransactionLib` for `approve` gasEstimation unpacking the flow
- recycle the upper `gasPrice` and use it in the `depositTokenAsync` with a fixed `gas`

Reply:
Done.

Reply:
I have used also `getNonce` of TransactionLib.